### ISSUE TITLE
Having a null in a multi index array avoids indexing

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4146]
User @eliaslevy ran into this.

```js
r.tableCreate('t1')
r.table('t1').insert({})
r.table('t1').indexCreate("test", function(x) { return ["a", null]; }, {multi: true})

r.table('t1').getAll("a", {index: "test"}).count()
-> 0
```

Having a `null` in the return value of a multi index function seems to stop all values in the array from being indexed. I would expect the row to still be indexed under `"a"`.

@Tryneus could you take a look?

**Comment by danielmewes** [comment]
Looking at the code we have `compute_keys` in btree.cc which loops over each key datum and calls `print_secondary` on it. `print_secondary` throws a ReQL exception if it encounters `null`, which aborts the whole process.
Might be easiest to just check for null before calling that, since we currently use `null` as a kind of special value that means "don't index this".

**Comment by mlucy** [comment]
I think we should change this so that any un-indexable values are skipped rather than special-casing `NULL`.  The easiest way to do this would be to catch the exception produced by `print_secondary` lower down and just continue on to the other fields.

**Comment by danielmewes** [comment]
@mlucy yeah that's actually what I'm now implementing. The other type of value that's relevant here are geometry objects, which would currently do the same as `null`.

**Comment by danielmewes** [comment]
Since this change can break a secondary index, moving to 2.1.

**Comment by danielmewes** [comment]
Fix in CR 2834

**Comment by danielmewes** [comment]
Fixed in `next` by e91e11f6f05050c3612b6f52ec35ddbebe19d875 . Will ship with RethinkDB 2.1.
